src/privatesend-util.h

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) 2014-2017 The Dash Core developers
+// Distributed under the MIT/X11 software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef PRIVATESENDUTIL_H
+#define PRIVATESENDUTIL_H
+
+#include "wallet/wallet.h"
+
+class KeyHolder
+{
+private:
+    CReserveKey reserveKey;
+    CPubKey pubKey;
+public:
+    KeyHolder(CWallet* pwalletIn);
+    void KeepKey();
+    void ReturnKey();
+
+    CScript GetScriptForDestination();
+
+};
+
+typedef std::shared_ptr<KeyHolder> KeyHolderPtr;
+
+class KeyHolderStorage
+{
+private:
+    std::vector<KeyHolderPtr> storage;
```

Comment:
Using `emplace_back` instead of `push_back` allows the key to be constructed in place, rather than constructed and copied:
`storage.emplace_back(KeyHolder(pwallet));`
(assuming a vector of KeyHolder)

I'm more concerned about avoiding lots of small allocations.
Storing objects, the vector manages the single 'elastic' allocation for all the keys.
Storing pointers, there is a new small allocations for each key added.